design/baremetal-operator/node-specific-kernel-args.md

```diff
@@ -0,0 +1,166 @@
+<!--
+ This work is licensed under a Creative Commons Attribution 3.0
+ Unported License.
+
+ http://creativecommons.org/licenses/by/3.0/legalcode
+-->
+
+# pre-provisioning kernel argument support
+
+## Status
+
+planned
+
+## Summary
+
+This document proposes a new BareMetalHost spec attribute intended for
+configuring the pre-provisioning image's (usually IPA) kernel arguments.
+
+## Motivation
+
+There are situations where a user would need node specific kernel arguments
+for the pre-provisioning image. The reason could be the use of a mixed set of
+hardware, troubleshooting provisioning or boot issues, selectively configuring
+custom logic e.g. enabling IPA plugins or other services embedded in the
+pre-provisioning image.
+
+### Goals
+
+1. Implement a string array type optional spec field named
+   `preProvisionKernelArgs`
+2. Extend the generic `Provisioner` interface to support the new optional field
+3. Implement support for the `preProvisionKernelArgs` in the Ironic provisioner
+
+### Non-Goals
+
+1. To implement any sort value check for the new optional spec
+2. Providing a default value for the new spec field
+
+## Proposal
+
+This document proposes a new BareMetalHost spec attribute intended for
+configuring the pre-provisioning image's (usually IPA) kernel arguments.
+
+The proposal involves the introduction of a new spec field thus a modification
+to the BMO API. The API change is intended to be backward compatible as the
+new `preProvisionKernelArgs` field intended to be optional.
+
+### Implementation Details/Notes/Constraints
+
+The `preProvisionKernelArgs` field would be located directly under `spec`
+section. The format of the new field would be a list/array of individual
+strings. The new field would be optional and it would be considered
+either a NOOP or a cleaning operation depending on the previous value of the
+field in case it would be not present or have no value or would have an empty
+array as a value.
+
+The proposed BMH looks like the following:
+
+```yaml
+apiVersion: metal3.io/v1alpha1
+kind: BareMetalHost
+metadata:
+ name: bm0
+spec:
+ online: true
+ bmc:
+   address: idrac://192.168.122.1:6230/
+   credentialsName: bm0-bmc-secret
+ preProvisionKernelArgs:
+   - --timeout
+   - 60000
+   - systemd.journald.forward_to_console=yes
+   - ipa-debug=1
+   ...
+   ...
+   ...
+ bootMACAddress: 52:54:00:b7:b2:6f
+...
+...
+...
+
+```
+
+The example contains a snippet of a common IPA kernel argument list.
+The `preProvisionKernelArgs` list would be received by the `Provisioner`
+interface as is. The handling of the list would be up to the particular
+implementations of the `Provisioner` interface. In case of the `Ironic`
+provisioner the elements from the `preProvisionKernelArgs` would be
+concatenated into a single string such that the string would start with the
```

Comment:
They are concatenated with spaces in between, right? I think that may be worth to point out explicitly here also.